### PR TITLE
Tidy up PropertiesLauncher a bit and clarify the behaviour

### PR DESCRIPTION
**Pull request by dsyer**
The main changes apart from documentation are

* Switch to `loader.properties` instead of `application.properties`
* Search for `loader.properties` in `loader.home` as well as in
the classpath
* Placeholder replacements in MANIFEST.MF (using `loader.properties`
or system/env vars)

Fixes gh-7221. I left in support for `application.properties`
but made `loader.properties` take precedence by default (we can
remove support later quite easily).

Also fixes gh-8334

### CODE REVIEW

**Comment by dsyer** [comment]
That test works for me.

**Comment by wilkinsona** [comment]
Thanks, @dsyer. Have I missed something or is the manifest still not used for some settings? It was point 6 on the list in #7221:

> The properties that can be customised using manifest entries seem rather arbitrary. For example `loader.config.name` and `loader.config.location` cannot be customised in this way.

**Comment by dsyer** [comment]
Correct. I didn't change that (the ones that are used are listed explicitly in the documentation already I think).

**Comment by wilkinsona** [comment]
They are, but only because I tried to make the documentation reflect the current situation in https://github.com/spring-projects/spring-boot/commit/ee7141cf63cf3dc59f038f666238682c81640187 which updated the second table in [this section](http://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#executable-jar-property-launcher-features). It looks like I missed `loader.config.name` though.

Arguably, the text prior to the first section is still misleading. Also, isn't it weird to only support some of the properties in the manifest? There's no reason why (almost?) all of them couldn't be supported is there?

**Comment by dsyer** [comment]
Not sure. I was thinking that since the properties can be used to locate the manifest, it doesn't make sense to allow it to contribute most of the properties really. I think we selected the ones that made sense (to me anyway).

**Comment by wilkinsona** [comment]
What about using the manifest to bake some customisation into an executable jar, for example by setting `Loader-Config-Name` or `Loader-Config-Location`? Seems like a reasonable thing to do to me.

**Comment by dsyer** [comment]
OK, I guess so. I think it's pretty dubious you'd get any value out of setting `Loader-Home` in the manifest. But I added support for all the properties in manifest in that last commit.

File: spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/PropertiesLauncher.java [hunk 1]
```diff
@@ -142,48 +140,61 @@ public PropertiesLauncher() {
...
-			}
-			finally {
-				resource.close();
+		List<String> configs = new ArrayList<String>();
+		if (getProperty(CONFIG_LOCATION) != null) {
+			configs.add(getProperty(CONFIG_LOCATION));
+		}
+		else {
+			// TODO: remove support for "application.properties"
+			String[] names = getPropertyWithDefault(CONFIG_NAME, "loader,application")
```
Comment (wilkinsona) [inline spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/PropertiesLauncher.java]: I think we should do this in such a way that we can log a warning that `application.properties` is deprecated and that `loader.properties` should be used instead. We can then remove support for `application.properties` in 2.0.

File: spring-boot-docs/src/main/asciidoc/appendix-executable-jar-format.adoc [hunk 2]
```diff
@@ -198,7 +198,13 @@ the appropriate launcher:
...
 `PropertiesLauncher` has a few special features that can be enabled with external
 properties (System properties, environment variables, manifest entries or
-`application.properties`).
+`loader.properties`).
+
+NOTE: `PropertiesLauncher` supports loading properties from
+`loader.properties` and also (for historic reasons)
+`application.properties`. To avoid confusion, we recommend using
+`loader.properties` exclusively, and support for
+`application.properties` may be removed in the future.
```
Comment (wilkinsona) [inline spring-boot-docs/src/main/asciidoc/appendix-executable-jar-format.adoc]: I think we should go further and say that support for `application.properties` is deprecated. We can then remove it in 2.0.

File: spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/PropertiesLauncher.java (hunk 1 above)
Comment (dsyer) [inline spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/PropertiesLauncher.java]: Done

File: spring-boot-docs/src/main/asciidoc/appendix-executable-jar-format.adoc (hunk 2 above)
Comment (dsyer) [inline spring-boot-docs/src/main/asciidoc/appendix-executable-jar-format.adoc]: Done

File: spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/PropertiesLauncher.java (hunk 1 above)
Comment (wilkinsona) [inline spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/PropertiesLauncher.java]: Thanks. Perhaps I've misread the change, but won't this warn if your app uses `application.properties` at all? For example, if there's one in `BOOT-INF/classes` in the jar that's configuring properties that don't affect `PropertiesLauncher`? I think we need to warn only if `application.properties` is found and it's used to configure one of the launcher's properties.

Comment (dsyer) [inline spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/PropertiesLauncher.java]: Good catch. Have another look now.

Comment (wilkinsona) [inline spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/PropertiesLauncher.java]: LGTM. Thanks

**Comment by dsyer** [comment]
Not easily no. Is it important?

**Comment by wilkinsona** [comment]
Yes, we think it is. I wouldn't have asked otherwise. They're separate changes so they belong in separate commits at the very least. Also, the changes for #7221 are an obvious candidate for 1.5.x, those for #8334 less so. Separate commits or, even better, separate PRs will make it much easier to process them.

**Comment by dsyer** [comment]
I actually would prefer all the changes go in 1.5 (why wait?). I'll see if it's possible to disentangle them.

**Comment by dsyer** [comment]
OK. It's 2 commits now.

**Comment by wilkinsona** [comment]
Thanks

**Pull request closed without merging**